src/dist/block_service/fds/fds_service.cpp

```diff
@@ -484,82 +498,139 @@ fds_file_object::fds_file_object(fds_service *s,
 
 fds_file_object::~fds_file_object() {}
 
-dsn::error_code fds_file_object::get_content(uint64_t pos,
-                                             int64_t length,
-                                             /*out*/ std::ostream &os,
-                                             /*out*/ uint64_t &transfered_bytes)
+error_code fds_file_object::get_file_meta()
+{
+    galaxy::fds::GalaxyFDSClient *c = _service->get_client();
+    try {
+        auto meta = c->getObjectMetadata(_service->get_bucket_name(), _fds_path)->metadata();
+        auto iter = meta.find(fds_service::FILE_LENGTH_CUSTOM_KEY);
+        if (iter != meta.end()) {
+            bool valid = dsn::buf2uint64(iter->second, _size);
+            if (!valid || _size < 0) {
+                derror("Error to get file size");
+                return ERR_FS_INTERNAL;
+            }
+        }
+
+        iter = meta.find(fds_service::FILE_MD5_KEY);
+        if (iter != meta.end()) {
+            _md5sum = iter->second;
+        }
+
+        _has_meta_synced = true;
+        return ERR_OK;
+    } catch (const galaxy::fds::GalaxyFDSClientException &ex) {
+        if (ex.code() == Poco::Net::HTTPResponse::HTTP_NOT_FOUND) {
+            return ERR_OBJECT_NOT_FOUND;
+        } else {
+            derror("fds getObjectMetadata failed: parameter(%s), code(%d), msg(%s)",
+                   _name.c_str(),
+                   ex.code(),
+                   ex.what());
+            return ERR_FS_INTERNAL;
+        }
+    }
+}
+
+error_code fds_file_object::get_content_in_batches(uint64_t start,
+                                                   int64_t length,
+                                                   /*out*/ std::ostream &os,
+                                                   /*out*/ uint64_t &transfered_bytes)
 {
-    dsn::error_code err;
+    // the max batch size is 1MB
+    const uint64_t BATCH_MAX = 1e6;
+    error_code err = ERR_OK;
     transfered_bytes = 0;
 
+    // get file meta if it is not synced
+    if (!_has_meta_synced) {
+        err = get_file_meta();
+        if (ERR_OK != err) {
+            return err;
+        }
+    }
+
+    // if length = -1, it means we should transfer the whole file
+    uint64_t to_transfer_bytes = length;
+    if (-1 == length) {
+        to_transfer_bytes = _size;
+    }
+
+    uint64_t pos = start;
+    uint64_t once_transfered_bytes = 0;
+    while (pos < start + to_transfer_bytes) {
+        uint64_t batch_len = std::min(BATCH_MAX, start + to_transfer_bytes - pos);
+        // get tokens from token bucket
+        _service->_token_bucket->consumeWithBorrowAndWait(batch_len);
+
+        err = get_content(pos, batch_len, os, once_transfered_bytes);
+        transfered_bytes += once_transfered_bytes;
+        if (err != ERR_OK || once_transfered_bytes < batch_len) {
+            return err;
+        }
+        pos += batch_len;
+    }
```

Comment:
Could you please explain when would the condition `pos >= start + to_transfer_bytes` happen to me? I just can not image when would meet such condition. Condition `err == ERR_OK && once_transfered_bytes < batch_size` means read the end of the file, and it return inside while.

Reply:
`pos == start + to_transfer_bytes` may happen, when `once_transfered_bytes` always equal with `batch_size`, and we read the end of the file.
but `pos > start + to_transfer_bytes` may not happy in normal case. But for the robustness of these code, I can use   `pos >= start + to_transfer_bytes` here.